--- include/linux/time.h-orig	2008-04-23 00:33:47.000000000 -0400
+++ include/linux/time.h	2008-04-23 00:34:23.000000000 -0400
@@ -169,7 +169,7 @@
  * @a:		pointer to timespec to be incremented
  * @ns:		unsigned nanoseconds value to be added
  */
-static inline void timespec_add_ns(struct timespec *a, u64 ns)
+static inline void timespec_add_ns(struct timespec *a, volatile u64 ns)
 {
 	ns += a->tv_nsec;
 	while(unlikely(ns >= NSEC_PER_SEC)) {

--- Makefile-orig	2008-04-23 00:39:54.000000000 -0400
+++ Makefile	2008-04-23 00:40:38.000000000 -0400
@@ -332,7 +332,8 @@
 
 KBUILD_CFLAGS   := -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs \
 		   -fno-strict-aliasing -fno-common \
-		   -Werror-implicit-function-declaration
+		   -Werror-implicit-function-declaration \
+                   -fno-tree-scev-cprop
 KBUILD_AFLAGS   := -D__ASSEMBLY__
 
 # Read KERNELRELEASE from include/config/kernel.release (if it exists)
